File: back/app/schemas/yougile.py
"""Схемы для эндпоинтов /api/yougile.

Главное правило: пароль и сам ключ YouGile в ответах НИКОГДА не показываем.
Наружу — только статус, last4 (fingerprint) и публичные id/названия выбранных
project/board/column.
"""
from marshmallow import Schema, fields, validate


# ── Подключение пользователя ───────────────────────────────────────────────

class YougileConnectStartSchema(Schema):
    """Получить список компаний по логину/паролю.

    Используется только на админском визарде — обычному пользователю list
    компаний не нужен, у него yg_company_id зафиксирован настройками.
    """
    login = fields.Email(required=True, validate=validate.Length(max=255))
    password = fields.String(required=True, load_only=True,
                             validate=validate.Length(min=1, max=512))


class YougileConnectFinishSchema(Schema):
    """Создать/перевыпустить ключ и сохранить привязку.

    Для обычного пользователя `yg_company_id` НЕ передаётся: бэк сам берёт
    его из настроек компании. Для админа в визарде — обязателен (он мог
    выбрать его в `/connect/start`).
    """
    login = fields.Email(required=True, validate=validate.Length(max=255))
    password = fields.String(required=True, load_only=True,
                             validate=validate.Length(min=1, max=512))
    yg_company_id = fields.String(load_default=None,
                                  validate=validate.Length(max=64))


class YougileRotateSchema(Schema):
    """Сброс ключа. Требуем повторно ввести пароль — это чувствительная
    операция, JWT-сессии недостаточно."""
    password = fields.String(required=True, load_only=True,
                             validate=validate.Length(min=1, max=512))


# ── Что показываем наружу ──────────────────────────────────────────────────

class YougileAccountStatusSchema(Schema):
    """Статус подключения текущего пользователя."""
    connected = fields.Boolean(required=True)
    yg_login = fields.String(allow_none=True)
    key_fingerprint = fields.String(allow_none=True)
    last_validated_at = fields.DateTime(allow_none=True)
    yg_company_id = fields.String(allow_none=True)
    # Признак «настроенной» интеграции компании — фронту удобно показать
    # фичу или баннер «попросите админа включить YouGile в компании».
    company_enabled = fields.Boolean(required=True)


class YougileCompanyListItemSchema(Schema):
    """Элемент списка `/auth/companies` — отдаём наружу только то, что нужно
    в UI визарда."""
    id = fields.String(required=True)
    name = fields.String(required=True)


class YougileProjectSchema(Schema):
    id = fields.String(required=True)
    title = fields.String(required=True)


class YougileBoardSchema(Schema):
    id = fields.String(required=True)
    title = fields.String(required=True)
    projectId = fields.String(allow_none=True)


class YougileColumnSchema(Schema):
    id = fields.String(required=True)
    title = fields.String(required=True)
    boardId = fields.String(allow_none=True)


# ── Настройки компании ────────────────────────────────────────────────────

class YougileCompanySettingsSchema(Schema):
    """Что отдаём админу: текущая конфигурация интеграции."""
    enabled = fields.Boolean(required=True)
    yg_company_id = fields.String(allow_none=True)
    yg_company_name = fields.String(allow_none=True)
    yg_project_id = fields.String(allow_none=True)
    yg_project_title = fields.String(allow_none=True)
    yg_board_id = fields.String(allow_none=True)
    yg_board_title = fields.String(allow_none=True)
    yg_first_column_id = fields.String(allow_none=True)
    yg_completed_column_id = fields.String(allow_none=True)
    webhook_registered = fields.Boolean(required=True)


class YougileImportTaskSchema(Schema):
    """Импорт карточки из YouGile в Groove Work.

    `url` — ссылка на карточку YG. Отдел всегда обязателен (как и при обычном
    создании задачи в GW). pull_deadline по умолчанию TRUE: подтянем дедлайн
    из YG-стикера, если он есть.
    """
    url = fields.String(required=True, validate=validate.Length(min=1, max=2000))
    department_id = fields.Integer(required=True)
    responsible_user_id = fields.Integer(load_default=None, allow_none=True)
    stage_id = fields.Integer(load_default=None, allow_none=True)
    pull_deadline = fields.Boolean(load_default=True)


class YougileExportTaskSchema(Schema):
    """Создать карточку в YouGile из существующей GW-задачи."""
    gw_task_id = fields.Integer(required=True)


class YougileCompanySettingsUpdateSchema(Schema):
    """Что админ может прислать. Все поля опциональны, чтобы можно было
    обновлять частями. enabled=False достаточно, чтобы выключить интеграцию,
    не теряя выбранные id."""
    enabled = fields.Boolean()
    yg_company_id = fields.String(validate=validate.Length(max=64), allow_none=True)
    yg_company_name = fields.String(validate=validate.Length(max=255), allow_none=True)
    yg_project_id = fields.String(validate=validate.Length(max=64), allow_none=True)
    yg_project_title = fields.String(validate=validate.Length(max=255), allow_none=True)
    yg_board_id = fields.String(validate=validate.Length(max=64), allow_none=True)
    yg_board_title = fields.String(validate=validate.Length(max=255), allow_none=True)
    yg_completed_column_id = fields.String(validate=validate.Length(max=64), allow_none=True)

File: back-go/tasks/internal/dto/yougile.go
package dto

// JSON-формы /api/yougile — байт-в-байт со схемами Flask
// (schemas/yougile.py); порядок полей алфавитный, как jsonify.

// YougileStatus — GET /api/yougile/status (YougileAccountStatusSchema).
type YougileStatus struct {
	CompanyEnabled  bool      `json:"company_enabled"`
	Connected       bool      `json:"connected"`
	KeyFingerprint  *string   `json:"key_fingerprint"`
	LastValidatedAt *JSONTime `json:"last_validated_at"`
	YgCompanyID     *string   `json:"yg_company_id"`
	YgLogin         *string   `json:"yg_login"`
}

// YougileConnectResult — POST /api/yougile/account.
type YougileConnectResult struct {
	Connected      bool   `json:"connected"`
	KeyFingerprint string `json:"key_fingerprint"`
	YgCompanyID    string `json:"yg_company_id"`
	YgLogin        string `json:"yg_login"`
}

// YougileRotateResult — POST /api/yougile/account/rotate.
type YougileRotateResult struct {
	Connected      bool   `json:"connected"`
	KeyFingerprint string `json:"key_fingerprint"`
}

// YougileCompanyItem — элемент POST /api/yougile/companies/lookup.
type YougileCompanyItem struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

// YougileProject / YougileBoard / YougileColumn — каталоги админ-визарда.
type YougileProject struct {
	ID    string `json:"id"`
	Title string `json:"title"`
}

type YougileBoard struct {
	ID        string  `json:"id"`
	ProjectID *string `json:"projectId"`
	Title     string  `json:"title"`
}

type YougileColumn struct {
	BoardID *string `json:"boardId"`
	ID      string  `json:"id"`
	Title   string  `json:"title"`
}

// YougileSettings — GET/PUT /api/yougile/company-settings и POST /reset
// (YougileCompanySettingsSchema).
type YougileSettings struct {
	Enabled             bool    `json:"enabled"`
	WebhookRegistered   bool    `json:"webhook_registered"`
	YgBoardID           *string `json:"yg_board_id"`
	YgBoardTitle        *string `json:"yg_board_title"`
	YgCompanyID         *string `json:"yg_company_id"`
	YgCompanyName       *string `json:"yg_company_name"`
	YgCompletedColumnID *string `json:"yg_completed_column_id"`
	YgFirstColumnID     *string `json:"yg_first_column_id"`
	YgProjectID         *string `json:"yg_project_id"`
	YgProjectTitle      *string `json:"yg_project_title"`
}

// ── Запросы (после schema-валидации в транспорте) ────────────────

// YougileConnect — POST /api/yougile/account (YougileConnectFinishSchema).
type YougileConnect struct {
	Login       string
	Password    string
	YgCompanyID *string
}

// YougileSettingsUpdate — PUT /company-settings: все поля опциональны,
// *Set = поле передано (частичное обновление, как "key in payload").
type YougileSettingsUpdate struct {
	Enabled                bool
	EnabledSet             bool
	YgCompanyID            *string
	YgCompanyIDSet         bool
	YgCompanyName          *string
	YgCompanyNameSet       bool
	YgProjectID            *string
	YgProjectIDSet         bool
	YgProjectTitle         *string
	YgProjectTitleSet      bool
	YgBoardID              *string
	YgBoardIDSet           bool
	YgBoardTitle           *string
	YgBoardTitleSet        bool
	YgCompletedColumnID    *string
	YgCompletedColumnIDSet bool
}

// YougileImport — POST /api/yougile/import-task (YougileImportTaskSchema).
type YougileImport struct {
	URL               string
	DepartmentID      int64
	ResponsibleUserID *int64
	StageID           *int64
	PullDeadline      bool
}
